x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,507 @@
+			s3Infos = append(s3Infos, s3Info{
+				region: record.AwsRegion,
+				name:   record.S3.bucket.Name,
+				key:    record.S3.object.Key,
+				arn:    record.S3.bucket.Arn,
+			})
+		}
+	}
+	return s3Infos, nil
+}
+
+func (p *Input) handleS3Objects(svc s3iface.ClientAPI, s3Infos []s3Info, errC chan error) error {
+	s3Context := &s3Context{
+		refs: 1,
+		errC: errC,
+	}
+
+	for _, s3Info := range s3Infos {
+		objectHash := s3ObjectHash(s3Info)
+
+		// read from s3 object
+		reader, err := newS3BucketReader(svc, s3Info, p.context)
+		if err != nil {
+			return errors.Wrap(err, "newS3BucketReader failed")
+		}
+
+		offset := 0
+		for {
+			log, err := reader.ReadString('\n')
+			if log == "" {
+				break
+			}
+
+			if err != nil {
+				if err == io.EOF {
+					// create event for last line
+					offset += len([]byte(log))
+					event := createEvent(log, offset, s3Info, objectHash, s3Context)
+					err = p.forwardEvent(event)
+					if err != nil {
+						err = errors.Wrapf(err, "forwardEvent failed for %v", s3Info.key)
+						s3Context.Fail(err)
+						return err
+					}
+					return nil
+				}
+				return errors.Wrapf(err, "ReadString failed for %v", s3Info.key)
+			}
+
+			// create event per log line
+			offset += len([]byte(log))
+			event := createEvent(log, offset, s3Info, objectHash, s3Context)
+			err = p.forwardEvent(event)
+			if err != nil {
+				err = errors.Wrapf(err, "forwardEvent failed for %v", s3Info.key)
+				s3Context.Fail(err)
+				return err
+			}
+		}
+	}
+
+	s3Context.done()
+	return nil
+}
+
+func newS3BucketReader(svc s3iface.ClientAPI, s3Info s3Info, context *channelContext) (*bufio.Reader, error) {
+	s3GetObjectInput := &s3.GetObjectInput{
+		Bucket: awssdk.String(s3Info.name),
+		Key:    awssdk.String(s3Info.key),
+	}
+	req := svc.GetObjectRequest(s3GetObjectInput)
+
+	resp, err := req.Send(context)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == awssdk.ErrCodeRequestCanceled {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "s3 get object request failed %v", s3Info.key)
+	}
+
+	return bufio.NewReader(resp.Body), nil
+}
+
+func (p *Input) forwardEvent(event beat.Event) error {
+	ok := p.outlet.OnEvent(event)
+	if !ok {
+		return errOutletClosed
+	}
+	return nil
+}
+
+func (p *Input) deleteMessage(queueURL string, messagesReceiptHandle string, svcSQS *sqs.Client) error {
+	deleteMessageInput := &sqs.DeleteMessageInput{
+		QueueUrl:      awssdk.String(queueURL),
+		ReceiptHandle: awssdk.String(messagesReceiptHandle),
+	}
+
+	req := svcSQS.DeleteMessageRequest(deleteMessageInput)
+	_, err := req.Send(p.context)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == awssdk.ErrCodeRequestCanceled {
+			return nil
+		}
+		return errors.Wrap(err, "DeleteMessageRequest failed")
+	}
+	return nil
+}
+
+func createEvent(log string, offset int, s3Info s3Info, objectHash string, s3Context *s3Context) beat.Event {
+	s3Context.Inc()
+	defer s3Context.done()
```

Comment:
Sorry to cause confusion here. Technically this call + defer on done is correct. It's the naive way. If it's known that ownership already exists + we have an `inc` dirrectly follow by a `done`, then we can optmize the `inc`+`done` away. This is only true, because we know that the caller to `createEvent` is in the same thread + the thread already has ownership (normally compilers do this work for you).

Reply:
I was overdoing it here... Removed!